README.md
![Skygear Logo](.github/skygear-logo.png)

# Python plugin runtime for Skygear

[![PyPI](https://img.shields.io/pypi/v/skygear.svg)](https://pypi.python.org/pypi/skygear)
[![Build Status](https://travis-ci.org/SkygearIO/py-skygear.svg)](https://travis-ci.org/SkygearIO/py-skygear)
[![License](https://img.shields.io/pypi/l/skygear.svg)](https://pypi.python.org/pypi/skygear)

When the Skygear Server calls your plugin, `py-skygear` will take the plugin message and calls the appropriate function automatically.

## Python Cloud Functions
`py-skygear` is useful for running Python Cloud Functions to work with [skygear-server](https://github.com/SkygearIO/skygear-server); `skygear-node` is the counter part of py-skygear but for Javascript.

You can read more about how to write Skygear Cloud Functions in [Skygear Guides](https://docs.skygear.io/guides/cloud-function/intro-and-deployment/python/). You don't need py-skygear if you're using skygear.io and don't plan to deploy it on your own.

## Install py-skygear

You may want to read [Setup Skygear Development Server Locally](https://docs.skygear.io/guides/advanced/server/) before installing py-skygear (and understand why you might or might not need `py-skygear`)

Install skygear by using pip. py-skygear requires Python 3.

```
$ pip3 install skygear
```

Alternatively, you can install `py-skygear` from source by cloning `py-skygear` from this official repository.

## Development

Skygear support two kind of protocol for different use case, make sure you
add support to all of them or raise appropriate exception.

Supported protocols: `http` and `zmq`

If you want to use zmq, you need to install pyzmq, and respective cbinding.
You can install via homebrew in OSX `brew install zeromq`

You may run a long running process that hook with your own skygear-server
instance.
```
DATABASE_URL=postgresql://localhost/skygear?sslmode=disable \
py-skygear sample.py \
--skygear-address tcp://127.0.0.1:5555 \
--skygear-endpoint http://127.0.0.1:3000 \
--apikey=API_KEY
```


## Support

For implementation related questions or technical support, please find us on the [official forum](https://discuss.skygear.io) or [community chat](https://slack.skygear.io).

If you believe you've found an issue with the Skygear Python cloud function SDK, please feel free to [report an issue](https://github.com/SkygearIO/py-skygear/issues).


## How to contribute

Pull Requests Welcome!

We really want to see Skygear grows and thrives in the open source community.
If you have any fixes or suggestions, simply send us a pull request!


## License & Copyright

```
Copyright (c) 2015-present, Oursky Ltd.
All rights reserved.

This source code is licensed under the Apache License version 2.0 
found in the LICENSE file in the root directory of this source tree. 
An additional grant of patent rights can be found in the PATENTS 
file in the same directory.

```


CHANGELOG.md
## 1.7.0 (2019-01-21)

### Other Notes

- Update requests>=2.20.0

## 1.6.1 (2018-09-04)

### Bug Fixes

- Fix registered provider being overwrite issue
- Fix missing offset for Query
- Fix percent escape asset name

## 1.6.0 (2018-06-26)

### Features

- Add Lambda support for Skygear Data Type #195
- Support structural logging with JSON formatter #207

### Bug Fixes

- Option should be optional when a default is specified
- Fix multiple registration of 'init' event
- Register multiple event handlers #209

### Other Notes

- Disallow registering for 'init' event
- Support `async_` hook param in addition to `async`
- Update to python 3.6 in travis and docker image

## 1.5.0 (2018-04-23)

### Other Notes

- Pubsub improvement, pubsub.publish supports multiple channel, sending the
    same data to multiple channels with the same connection
- Fix unable to change record ACL settings in before save

## 1.4.1 (2018-04-09)

### Other Notes

- Update settings parsing (#186, #187)
- Update bcrypt to 3.1.4


## 1.4.0 (2018-03-07)

### Other Notes

- Update psycopg2 for supporting postgres10.1
- Update README for removing exec transport

## 1.3.0 (2018-01-04)

### Bug Fixes 

- Fix register handler with different HTTP method (#164)

## 1.2.0 (2017-12-07)

### Features

- Initial Database Query Interface Implementation

### Other Notes

- Error in skygear response should not raise Exception in http as in zmq

## 1.1.1 (2017-11-13)

### Features

- Inject version to request (#11, #153)
- Reflect views in utils.db (SkygearIO/forgot_password#23)

### Other Notes

- Update README.md for broken links and clarity
- Fix Make files
- Update example

## 1.1.0 (2017-08-07)

### Bug Fixes

- Sign FS Asset URL correctly (skygeario/skygear-server#427)

### Other Notes

- Update the desc on HTTP_ADDR option

## 1.0.0 (2017-06-30)

No changes since last release.

## 0.24.0 (2017-05-23)

### Features

- Make plugin transport ZMQ multiplex (SkygearIO/skygear-server#295)

    This feature will enable the bidirectional plugin transport and multiplexing.

    Calling `send_action` within the plugin will route back to the same worker
    from now on. It is implicit, developer does not need to specific worker reuse
    or not.
    This merge also add dynamic threading at ZMQTransport.
    It solved the problem of workers exhaust problem.

### Bug Fixes

- Fix unable to set envvar to `false` (#137)

## 0.23.0 (2017-04-20)

### Features

- Add user friendly error message to error class SkygearIO/skygear-SDK-Android#89

## 0.22.2 (2017-03-31)

### Features
- Read content type in asset serialisation SkygearIO/skygear-SDK-JS#164

### Bug Fixes
- Update the mock interface for ZMQTransport (#122)

### Other Notes
- Add back the missing boto3 deps at setup.py


## 0.22.0 (2017-02-10)

### Features

- Add timeout parameter to send_action (skygeario/skygear-server#271)

### Bug Fixes

- Fix passing key_required to server (#117)

## 0.21.0 (2017-01-11)

### Features

- Enable sending push notification by topic (SkygearIO/skygear-server#239)

### Other Notes

- Add documentation of `includeme`

## 0.20.0 (2016-12-20)

### Features

- Support s3 asset store url prefix
- Enable zmq thread pool configuration (#76)
- Support asset url signer (#105)
- Add support for UnknownValue type (SkygearIO/skygear-server#231, #101)

### Bug Fixes

- Fix postgis-2.3 not found
- Config module from directory once is loaded if it is specifid in
  options.modules (#70)
- Auto fix static assets prefix if slash is specified (#95)

### Incompatible changes

- Remove `skygear.skyconfig`

## 0.19.0 (2016-11-10)

### Incompatible changes

- The protocol for plugin transport is updated. Skygear Server and cloud code
  in previous versions cannot be used with this version.

### Features

- Support for plugin only request during initialization (SkygearIO/skygear-server#219)
- Support plugin event (#79)
- Add ignore-public-html option

### Bug Fixes

- Fix zmq stopping bug on maintains_worker_count (#80)

### Other Notes

- Fix various issues with docker compose and Makefile
- Handle http plugin request encoded in request body (#82)
- Check opened worker count instead of current active worker

## 0.18.0 (2016-10-28)

### Features

- Make the zmq transmitter open worker thread (#72)
- Provide skygear.config for loading and config of plugin (#70)
- Add settings utility (#62)
- Override previously defined extension points (#58)

### Other Notes

- Push Docker image as `onbuild` also (#66)
- Update to use py3.5 as default release version (#63)

## 0.17.0 (2016-09-15)

### Features

- Load more than one cloudcode entry points (#57)

## 0.16.0 (2016-09-02)

### Bug Fixes

- Fix local static asset not serve correctly during development (#50)

### Other Notes

- Un-deprecate PUBSUB_URL (#52)
- Revamp release process

## 0.15.0 (2016-08-17)

### Other Notes

- Update postgres to 9.5 (#45)
- Update catapi docker-compose to embrace env var config
- Remove the SKYGEAR prefix in envvar (#27)

## 0.14.0 (2016-07-26)

### Features

- send_action support datetime serialization using rfc3339 (#12)
- Add exception_handler decorator(#19)
- Register citext in postgresql dialects (#31)

## 0.13.1 (2016-07-08)

### Features
- Do not serve static assets by default

## 0.13.0 (2016-07-05)

### Features

- Include both api_key and access_token in send_action
- Provide meta attributes like created_by on serialized_record (SkygearIO/chat#23)
- Allow RestfulRecord subclass to provide container
- Implement serving static assets for development (#23)
- Automatically generate ID for restful record
- Enable threading in http transport
- Add function to check if a table exists in schema
- Declare and collect static assets (#21, #22)

### Bug fixes

- Fix handler name with slashes in http transport (#33, oursky/py-skygear#137)
- Fix handler returning werkzeug response

### Other notes

- Fix bcrypt version at 2.0.0 because build failing
- Fix onbuild image not having zmq transport installed


skygear/error.py
# Copyright 2015 Oursky Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
NotAuthenticated = 101
PermissionDenied = 102
AccessKeyNotAccepted = 103
AccessTokenNotAccepted = 104
InvalidCredentials = 105
InvalidSignature = 106
BadRequest = 107
InvalidArgument = 108
Duplicated = 109
ResourceNotFound = 110
NotSupported = 111
NotImplemented = 112
ConstraintViolated = 113
IncompatibleSchema = 114
AtomicOperationFailure = 115
PartialOperationFailure = 116
UndefinedOperation = 117
PluginUnavailable = 118
PluginTimeout = 119
RecordQueryInvalid = 120
PluginInitializing = 121

UnexpectedError = 10000


class SkygearException(Exception):
    """
    An exception object to be raised by plugin to indicate that
    the requested operation is not successful, or that the operation
    is aborted for whatever reason.
    """
    def __init__(self, message, code=UnexpectedError, info=None):
        self.message = message
        self.code = code
        self.info = info or {}
        super(SkygearException, self).__init__(message)

    def as_dict(self):
        return {
            'message': self.message,
            'code': self.code,
            'info': self.info,
            }

    def readable_message(self):
        """
        This method returns a user readable string for the error.
        """
        # Make exception of the complexity check here because we need to
        # cover each error case.
        # pylama:ignore=C901
        # #lizard forgives the complexity
        if self.code == NotAuthenticated:
            return "You have to be authenticated to perform this operation."
        elif (self.code in [PermissionDenied, AccessKeyNotAccepted,
                            AccessTokenNotAccepted]):
            return "You are not allowed to perform this operation."
        elif self.code == InvalidCredentials:
            return "You are not allowed to log in because the credentials " + \
                "you provided are not valid."
        elif self.code in [InvalidSignature, BadRequest]:
            return "The server is unable to process the request."
        elif self.code == InvalidArgument:
            return "The server is unable to process the data."
        elif self.code == Duplicated:
            return "This request contains duplicate of an existing " + \
                "resource on the server."
        elif self.code == ResourceNotFound:
            return "The requested resource is not found."
        elif self.code == NotSupported:
            return "This operation is not supported."
        elif self.code == NotImplemented:
            return "This operation is not implemented."
        elif (self.code in [ConstraintViolated, IncompatibleSchema,
                            AtomicOperationFailure, PartialOperationFailure]):
            return "A problem occurred while processing this request."
        elif self.code == UndefinedOperation:
            return "The requested operation is not available."
        elif self.code in [PluginInitializing, PluginUnavailable]:
            return "The server is not ready yet."
        elif self.code == PluginTimeout:
            return "The server took too long to process."
        elif self.code == RecordQueryInvalid:
            return "A problem occurred while processing this request."
        return "An unexpected error has occurred."

    @classmethod
    def from_dict(cls, error_dict):
        message = error_dict.get('message', "An error has occurred.")
        code = error_dict.get('code', UnexpectedError)
        info = error_dict.get('info', {})
        return cls(message, code, info)
